File: ml/scripts/run_server.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
EngageAI ML Service Server

This script starts the FastAPI server for the ML service.

Usage:
    python run_server.py [--host HOST] [--port PORT] [--reload]
"""

import os
import sys
import argparse
import logging
import uvicorn
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(),
        logging.FileHandler(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 
                                         'logs', 'server.log'))
    ]
)

logger = logging.getLogger(__name__)

# Ensure logs directory exists
os.makedirs(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'logs'), exist_ok=True)

def parse_args():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(description='Run the EngageAI ML service server')
    
    parser.add_argument('--host', type=str, 
                        default=os.getenv('HOST', '0.0.0.0'),
                        help='Host to run the server on')
    parser.add_argument('--port', type=int, 
                        default=int(os.getenv('PORT', 8000)),
                        help='Port to run the server on')
    parser.add_argument('--reload', action='store_true',
                        help='Enable auto-reload for development')
    
    return parser.parse_args()

def main():
    """Main function to run the server."""
    args = parse_args()
    
    logger.info(f"Starting EngageAI ML service server on {args.host}:{args.port}")
    
    # Add the project root to the Python path
    import sys
    import os
    sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    
    # Run the server
    uvicorn.run(
        "src.main:app",
        host=args.host,
        port=args.port,
        reload=args.reload,
        log_level="info"
    )

if __name__ == "__main__":
    main()

File: backend/scripts/run_server.js
#!/usr/bin/env node

/**
 * EngageAI Backend Server
 * 
 * This script starts the Express server for the backend service.
 * 
 * Usage:
 *   node run_server.js [--port PORT]
 */

require('dotenv').config();
const { program } = require('commander');
const path = require('path');
const fs = require('fs');
const { spawn } = require('child_process');

// Parse command line arguments
program
  .option('-p, --port <port>', 'Port to run the server on', process.env.PORT || 5000)
  .option('-e, --env <env>', 'Environment to run in', process.env.NODE_ENV || 'development')
  .parse(process.argv);

const options = program.opts();

// Ensure logs directory exists
const logsDir = path.join(__dirname, '..', 'logs');
if (!fs.existsSync(logsDir)) {
  fs.mkdirSync(logsDir, { recursive: true });
}

// Create log file streams
const accessLogStream = fs.createWriteStream(
  path.join(logsDir, 'access.log'),
  { flags: 'a' }
);

const errorLogStream = fs.createWriteStream(
  path.join(logsDir, 'error.log'),
  { flags: 'a' }
);

// Set environment variables
process.env.PORT = options.port;
process.env.NODE_ENV = options.env;

console.log(`Starting EngageAI backend server in ${options.env} mode on port ${options.port}`);

// Start the server
const serverProcess = spawn('node', ['src/index.js'], {
  stdio: ['inherit', 'pipe', 'pipe'],
  env: process.env
});

// Pipe stdout and stderr to console and log files
serverProcess.stdout.pipe(process.stdout);
serverProcess.stderr.pipe(process.stderr);
serverProcess.stdout.pipe(accessLogStream);
serverProcess.stderr.pipe(errorLogStream);

// Handle server process events
serverProcess.on('error', (error) => {
  console.error(`Failed to start server: ${error.message}`);
  process.exit(1);
});

serverProcess.on('exit', (code, signal) => {
  if (code !== 0) {
    console.error(`Server process exited with code ${code} and signal ${signal}`);
    process.exit(code);
  }
});

// Handle process signals
process.on('SIGINT', () => {
  console.log('Received SIGINT. Shutting down gracefully...');
  serverProcess.kill('SIGINT');
});

process.on('SIGTERM', () => {
  console.log('Received SIGTERM. Shutting down gracefully...');
  serverProcess.kill('SIGTERM');
});

process.on('uncaughtException', (error) => {
  console.error(`Uncaught exception: ${error.message}`);
  serverProcess.kill('SIGTERM');
  process.exit(1);
});